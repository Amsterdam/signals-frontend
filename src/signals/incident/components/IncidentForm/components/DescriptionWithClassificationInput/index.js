import React from 'react';
import PropTypes from 'prop-types';

import Title from '../Title/';
import ErrorMessage from '../ErrorMessage/';

const DescriptionWithClassificationInput = ({ handler, touched, hasError, meta, parent, getError }) => (
  <div>
    {meta.ifVisible ?
      <div className={`row mode_input verplicht ${touched && hasError('required') ? 'row_ongeldig' : ''}`}>
        <Title meta={meta} />
        <div className="input-help col-12">
          <ErrorMessage
            touched={touched}
            hasError={hasError}
            getError={getError}
          />
        </div>
        <div className={`col-${meta.cols || 12} invoer`}>
          <textarea
            className="input"
            rows={meta.rows || 6}
            placeholder={meta.placeholder}
            {...handler()}
            onKeyUp={(e) => parent.meta.getClassification(e.target.value)}
          />
        </div>
      </div>
       : ''}
  </div>
);

DescriptionWithClassificationInput.propTypes = {
  handler: PropTypes.func.isRequired,
  touched: PropTypes.bool.isRequired,
  hasError: PropTypes.func.isRequired,
  meta: PropTypes.object.isRequired,
  parent: PropTypes.object.isRequired,
  getError: PropTypes.func.isRequired,
};

export default DescriptionWithClassificationInput;
