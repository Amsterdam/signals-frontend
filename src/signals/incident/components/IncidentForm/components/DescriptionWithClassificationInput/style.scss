.text-area-input {
    &__counter {
      margin-left: 5px;
    }
  }
  