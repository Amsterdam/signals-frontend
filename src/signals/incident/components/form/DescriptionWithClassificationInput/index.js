import React from 'react';
import PropTypes from 'prop-types';

import Header from '../Header/';

import './style.scss';

function get(e, meta, parent) {
  parent.meta.getClassification(e.target.value);
  parent.meta.updateIncident({ [meta.name]: e.target.value });
}


const DescriptionWithClassificationInput = ({ handler, touched, value, hasError, meta, parent, getError, validatorsOrOpts }) => (
  <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
    {meta && meta.isVisible ?
      <div className={`${meta.className || 'col-12'} mode_input`}>
        <Header
          meta={meta}
          options={validatorsOrOpts}
          touched={touched}
          hasError={hasError}
          getError={getError}
        >
          <div className="invoer">
            <textarea
              rows={meta.rows || 6}
              placeholder={meta.placeholder}
              {...handler()}
              onBlur={(e) => get(e, meta, parent)}
            />
            { meta.maxLength &&
              <div className="input-help">
                <span className="text-area-input__counter">
                  {`${value ? value.length : '0'}/${meta.maxLength} tekens` }
                </span>
              </div>
            }
          </div>
        </Header>
      </div>
       : ''}
  </div>
);

DescriptionWithClassificationInput.propTypes = {
  handler: PropTypes.func,
  touched: PropTypes.bool,
  value: PropTypes.string,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func,
  validatorsOrOpts: PropTypes.object
};

export default DescriptionWithClassificationInput;
