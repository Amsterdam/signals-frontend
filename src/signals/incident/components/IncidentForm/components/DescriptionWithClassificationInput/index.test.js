import React from 'react';
import { shallow } from 'enzyme';

import DescriptionWithClassificationInput from './index';

describe('Form component <DescriptionWithClassificationInput />', () => {
  const metaFields = {
    name: 'input-field-name',
    placeholder: 'type here'
  };
  let wrapper;
  let handler;
  let touched;
  let getError;
  let hasError;
  let parent;

  beforeEach(() => {
    handler = jest.fn();
    touched = false;
    getError = jest.fn();
    hasError = jest.fn();
    parent = {
      meta: {
        updateIncident: jest.fn(),
        getClassification: jest.fn()
      }
    };

    wrapper = shallow(<DescriptionWithClassificationInput
      handler={handler}
      parent={parent}
      touched={touched}
      hasError={hasError}
      getError={getError}
    />);
  });

  describe('rendering', () => {
    it('should render classification text area field correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      expect(handler).toHaveBeenCalledWith();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no classification text area field when not visible', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: false
        }
      });

      expect(handler).not.toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render with a character counter when maxLength is supplied correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          maxLength: 3000
        }
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with a character counter with value correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
          maxLength: 3000
        },
        value: 'test'
      });

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    const event = { target: { value: 'diabolo' } };

    it('sets incident when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true
        }
      });

      wrapper.find('textarea').simulate('blur', event);

      expect(parent.meta.getClassification).toHaveBeenCalledWith('diabolo');
      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        'input-field-name': 'diabolo'
      });
    });
  });
});
