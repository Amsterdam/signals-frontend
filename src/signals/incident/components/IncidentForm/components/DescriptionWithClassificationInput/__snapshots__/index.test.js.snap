// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <DescriptionWithClassificationInput /> rendering should render classification text area field correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "name": "input-field-name",
          "placeholder": "type here",
        }
      }
      touched={false}
    >
      <div
        className="invoer"
      >
        <textarea
          onBlur={[Function]}
          placeholder="type here"
          rows={6}
        />
      </div>
    </Header>
  </div>
</div>
`;

exports[`Form component <DescriptionWithClassificationInput /> rendering should render no classification text area field when not visible 1`] = `
<div
  className=""
/>
`;

exports[`Form component <DescriptionWithClassificationInput /> rendering should render with a character counter when maxLength is supplied correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "maxLength": 3000,
          "name": "input-field-name",
          "placeholder": "type here",
        }
      }
      touched={false}
    >
      <div
        className="invoer"
      >
        <textarea
          onBlur={[Function]}
          placeholder="type here"
          rows={6}
        />
        <div
          className="input-help"
        >
          <span
            className="text-area-input__counter"
          >
            0/3000 tekens
          </span>
        </div>
      </div>
    </Header>
  </div>
</div>
`;

exports[`Form component <DescriptionWithClassificationInput /> rendering should render with a character counter with value correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-12 mode_input"
  >
    <Header
      getError={[MockFunction]}
      hasError={[MockFunction]}
      meta={
        Object {
          "isVisible": true,
          "maxLength": 3000,
          "name": "input-field-name",
          "placeholder": "type here",
        }
      }
      touched={false}
    >
      <div
        className="invoer"
      >
        <textarea
          onBlur={[Function]}
          placeholder="type here"
          rows={6}
        />
        <div
          className="input-help"
        >
          <span
            className="text-area-input__counter"
          >
            4/3000 tekens
          </span>
        </div>
      </div>
    </Header>
  </div>
</div>
`;
